pkg/controllers/cluster/cluster_controller.go

```diff
@@ -194,3 +248,155 @@ func (c *Controller) createExecutionSpace(cluster *v1alpha1.Cluster) error {
 
 	return nil
 }
+
+func (c *Controller) monitorClusterHealth() error {
+	clusterList := &v1alpha1.ClusterList{}
+	if err := c.Client.List(context.TODO(), clusterList); err != nil {
+		return err
+	}
+
+	clusters := clusterList.Items
+	for i := range clusters {
+		cluster := &clusters[i]
+		if err := wait.PollImmediate(MonitorRetrySleepTime, MonitorRetrySleepTime*HealthUpdateRetry, func() (bool, error) {
+			// Cluster object may be changed in this function.
+			err := c.tryUpdateClusterHealth(cluster)
+			if err == nil {
+				return true, nil
+			}
+			clusterName := cluster.Name
+			clusterObj := &v1alpha1.Cluster{}
+			if err := c.Get(context.TODO(), client.ObjectKey{Name: clusterName}, clusterObj); err != nil {
+				// If the cluster does not exist any more, we delete the health data from the map.
+				if apierrors.IsNotFound(err) {
+					c.clusterHealthMap.Delete(clusterName)
+					return true, nil
+				}
+				klog.Errorf("Getting a cluster to retry updating cluster health error: %v", clusterName, err)
+				return false, err
+			}
+			return false, nil
+		}); err != nil {
+			klog.Errorf("Update health of Cluster '%v' from Controller error: %v. Skipping.", cluster.Name, err)
+			continue
+		}
+	}
+
+	return nil
+}
+
+// tryUpdateClusterHealth checks a given cluster's conditions and tries to update it.
+func (c *Controller) tryUpdateClusterHealth(cluster *v1alpha1.Cluster) error {
```

Comment:
I know it's similar to k/k's node life cycle controller, but this method is very hard to understand, don't you think that?

The `cyclomatic complexities` is too high~, let's think about how to improve it.

Reply:
Sure, I'm looking forward.